versus/js/05-asynchronous-computation/os-events-description.html

```diff
@@ -0,0 +1,59 @@
+	In this role the operating system (OS) is a service provider which exposes blocking and non-blocking methods.
+	Delegating thread management and polling loops to the OS by using non-blocking asynchronous OS methods,
+	the C++ program is reduced in its complexity and is made less error-prone without being unresponsive during
+	the execution time of background computations.
+</p>
+<p>
+	Asynchrony can also be achieved with <em>long-running internal computations</em> by moving them into background threads.
+	To avoid blocking the main thread of execution the program instructs the OS to execute the background thread and signal
+	the main thread when the background thread has exited.
+</p>
+<p>
+	Calling non-blocking OS methods from C++ is trivial (it's what C++ is made for), but to get notified of external events, once an
+	event loop must be implemented and registered with the OS. Also OS-dependent C-callbacks or background threads
+	have to be registered for each non-blocking method.
+</p>
+<p>
+	<a href="https://www.boost.org/doc/libs/1_68_0/doc/html/boost_asio.html">Boost.Asio</a>
+	provides a platform-independent asynchronous api for common long-running tasks in networking, timers, serial ports and signals.
+	The outcome events are typically processed by continuations (callback functions) or can be forwarded to <code>std::future</code>.
+</p>
+<p>
+	<a href="https://www.boost.org/doc/libs/1_68_0/doc/html/thread.html">Boost.Thread</a> v.4 provides a implementation for
+	future continuations as proposed in <a href="http://www.modernescpp.com/index.php/component/content/article/44-blog/multithreading/multithreading-c-17-and-c-20/279-std-future-extensions?Itemid=239">C++14 TS Extensions for Concurrency V1</a>. C++ Futures are a general high-level abstraction for background threads.
+	With the <code>.then</code> continuations they have much in common with JS promises.
+</p>
+<h4>What's the Same</h4>
+<p>
+	In both cases the code sets up an environment to support asynchronous execution. Both programs run non-blocking down to
+	the bottom of the program without any output.
+
+	Two examples of asynchronous work are given:
+	<ul>
+		<li>Scheduling a Hello-World-lambda for immediate deferred execution</li>
+		<li>Quitting the program on an external <code>KILL</code> signal</li>
+	</ul>
+</p>
+<h4>What's Different</h4>
+<p>
+	<h5>Node.js</h5>
+		JavaScript can rely on the engine for all OS interfacing stuff. It does not need extra libraries.
+		The intricacies of the Node.js event loop are described <a href="https://nodejs.org/en/docs/guides/event-loop-timers-and-nexttick/">here</a>.
+
+	<h5>Boost.Asio</h5>
+	The boilerplate code to setup the <code>Boost.Asio</code> global context is minimal.
+	<p>
+		Boost.Asio commences all background work in a single method <code>asio::context::run</code>.
+		The method blocks as long as there are pending asynchronous computations. In the other event
+		the program will quit.
```

Comment:
I don't think this last sentence is needed. The next paragraph explains it better.